File: mobile/modules/core/android/src/main/java/com/mentra/core/ObservableStore.kt
package com.mentra.core

import org.json.JSONObject

/** Observable state management with immediate event emission */
class ObservableStore {
    private val values = mutableMapOf<String, Any>()
    private var onEmit: ((String, Map<String, Any>) -> Unit)? = null

    fun configure(onEmit: (String, Map<String, Any>) -> Unit) {
        this.onEmit = onEmit
    }

    fun set(category: String, key: String, value: Any) {
        val fullKey = "$category.$key"
        val oldValue = values[fullKey]

        // Skip if unchanged
        if (oldValue != null && toJson(oldValue) == toJson(value)) return

        values[fullKey] = value

        // Emit immediately
        onEmit?.invoke(category, mapOf(key to value))
    }

    fun get(category: String, key: String): Any? = values["$category.$key"]

    fun getCategory(category: String): Map<String, Any> {
        val prefix = "$category."
        return values.filterKeys { it.startsWith(prefix) }.mapKeys { it.key.removePrefix(prefix) }
    }

    private fun toJson(value: Any): String {
        return JSONObject(mapOf("v" to value)).toString()
    }
}

File: mobile/modules/core/ios/Source/ObservableStore.swift
//
//  ObservableStore.swift
//  Core
//
//  Observable state management with immediate event emission
//

import Foundation

@MainActor
class ObservableStore {
    private nonisolated(unsafe) var values: [String: Any] = [:]
    private var onEmit: ((String, [String: Any]) -> Void)?

    func configure(onEmit: @escaping (String, [String: Any]) -> Void) {
        self.onEmit = onEmit
    }

    func set(_ category: String, _ key: String, _ value: Any) {
        let fullKey = "\(category).\(key)"
        let oldValue = values[fullKey]

        // Skip if unchanged
        if let old = oldValue, areEqual(old, value) {
            return
        }

        values[fullKey] = value

        // Emit immediately
        onEmit?(category, [key: value])
    }

    nonisolated func get(_ category: String, _ key: String) -> Any? {
        values["\(category).\(key)"]
    }

    func getCategory(_ category: String) -> [String: Any] {
        var result: [String: Any] = [:]
        let prefix = "\(category)."
        for (key, value) in values where key.hasPrefix(prefix) {
            let shortKey = String(key.dropFirst(prefix.count))
            result[shortKey] = value
        }
        return result
    }

    // Helper to compare values
    private func areEqual(_ lhs: Any, _ rhs: Any) -> Bool {
        if let l = lhs as? String, let r = rhs as? String { return l == r }
        if let l = lhs as? Int, let r = rhs as? Int { return l == r }
        if let l = lhs as? Bool, let r = rhs as? Bool { return l == r }
        if let l = lhs as? Double, let r = rhs as? Double { return l == r }
        if let l = lhs as? [String], let r = rhs as? [String] { return l == r }
        if let l = lhs as? [[String: Any]], let r = rhs as? [[String: Any]] {
            return toJson(l) == toJson(r)
        }
        return false
    }

    private func toJson(_ value: Any) -> String? {
        guard let data = try? JSONSerialization.data(withJSONObject: value) else { return nil }
        return String(data: data, encoding: .utf8)
    }
}
